engine/linksystem.go

```diff
@@ -83,30 +83,33 @@ func (e *Engine) mkLinkSystem() ipld.LinkSystem {
 			// If the link is not found, it means that the root link of the list has
 			// not been generated and we need to get the relationship between the cid
 			// received and the contextID so the lister knows how to
-			// regenerate the list of CIDs.
-			key, err := e.getCidKeyMap(ctx, c)
+			// regenerate the list of CIDs. There can be multiple provider + context IDs associated with a cid
+			// as index provider supports publishing on behalf of multiple providers.
+			keys, err := e.getAllCidKeyMap(ctx, c)
 			if err != nil {
 				log.Errorf("Error fetching relationship between CID and contextID: %s", err)
 				return nil, err
 			}
 
-			// Get the car iterator needed to create the entry chunks.
-			// Normally for removal this is not needed since the indexer
-			// deletes all indexes for the contextID in the removal
-			// advertisement.  Only if the removal had no contextID would the
-			// indexer ask for entry chunks to remove.
-			mhIter, err := e.mhLister(ctx, key)
-			if err != nil {
-				return nil, err
-			}
-
-			// Store the linked list entries in cache as we generate them.  We
-			// use the cache linksystem that stores entries in an in-memory
-			// datastore.
-			_, err = e.entriesChunker.Chunk(ctx, mhIter)
-			if err != nil {
-				log.Errorf("Error generating linked list from multihash lister: %s", err)
-				return nil, err
+			for _, key := range keys {
```

Comment:
If the CID for a set of entries is the the same, then by definition it means the multihashes they contained are also the same, since CID contains the digest of blocks it is associated to (with some caveats around hash function used) right?

If so then why entires are chunked for each of the `proivderandcontext` keys? 
Would doing this for any one of them once not be sufficient?

Reply:
Really great point, thanks! What that got me into thinking - is that we don't need to store cid -> context mappings for *all* providers. Storing it for *any* provider could be enough as chunking it for just one of them would be sufficient. Changed the code to maintain just one mapping and got rid of the expensive scan.